Restructure GDB debugging and add guide for STLink and link to emdbg (#2651)
PX4/PX4-user_guide: en/debug/gdb_debugging.md

# Debugging with GDB

The [GNU DeBugger (GDB)](https://sourceware.org/gdb/download/onlinedocs/gdb/index.html) comes installed with the compiler toolchain in the form of the `arm-none-eabi-gdb` binary.
The debugger reads the debug symbols inside an ELF file to understand the static and dynamic memory layout of the PX4 firmware.

> To access the PX4 autopilot microcontroller, it needs to connect to a [Remote Target](https://sourceware.org/gdb/download/onlinedocs/gdb/Connecting.html), which is provided by a [SWD debug probe](swd_debug).

Reviewer: ```suggestion
To access the PX4 autopilot microcontroller, it needs to connect to a [Remote Target](https://sourceware.org/gdb/download/onlinedocs/gdb/Connecting.html), which is provided by a [SWD debug probe](swd_debug.md).
```

The flow of information looks like this:

```
Developer <=> GDB <=> GDB Server <=> Debug Probe <=> SWD <=> PX4 Autopilot.
```

## Quickstart

To start a debugging session you typically:

> 1. Need a specialized [SWD debug probe](swd_debug#debug-probes).

Reviewer: The link works but the anchor doesn't - you've pushed all the probes into their own docs.  
1. We probably need a list of probes in that page, each linking to the actual probe - right?
2. The page does have debug adaptors. Should we pull them out as peers of the probes too?

Reviewer: ```suggestion
1. Need a specialized [SWD debug probe](../debug/swd_debug.md#debug-probes).
```

> 2. Find and connect to the [SWD debug port](swd_debug#debug-signals), [you may need an adapter](swd_debug#debug-probes).

Reviewer: ```suggestion
2. Find and connect to the [SWD debug port](swd_debug#debug-signals), [you may need an adapter](swd_debug.md#debug-probes).
```

> 2. Find and connect to the [SWD debug port](swd_debug#debug-signals), [you may need an adapter](swd_debug.md#debug-probes).

Reviewer: ```suggestion
2. Find and connect to the [SWD debug port](../debug/swd_debug.md#autopilot-debug-ports).
   You may need a [debug adapter](swd_debug.md#debug-adapter).
```

3. Configure and start the debug probe to create a GDB server.
4. Launch GDB and connect to the GDB server as a remote target.
5. Debug your firmware interactively.

See the debug probe documentation for details on how to setup your debug connection:

> - [SEGGER J-Link](probe_jlink): commercial probe, no built-in serial console, requires adapter.

Reviewer: ```suggestion
- [SEGGER J-Link](probe_jlink.md): commercial probe, no built-in serial console, requires adapter.
```

> - [Black Magic Probe](probe_bmp): integrated GDB server and serial console, requires adapter.

Reviewer: ```suggestion
- [Black Magic Probe](probe_bmp.md): integrated GDB server and serial console, requires adapter.
```
